lib/SILOptimizer/IPO/GlobalOpt.cpp

```diff
@@ -955,27 +1026,72 @@ bool SILGlobalOpt::run() {
           continue;
         }
 
-        auto *GAI = dyn_cast<GlobalAddrInst>(&I);
-        if (!GAI) {
+        if (auto *GAI = dyn_cast<GlobalAddrInst>(&I)) {
+          collectGlobalAccess(GAI);
           continue;
         }
 
-        collectGlobalAccess(GAI);
+        if (auto *allocGlobal = dyn_cast<AllocGlobalInst>(&I)) {
+          collectAllocGlobal(allocGlobal->getReferencedGlobal(), allocGlobal);
+          continue;
+        }
       }
     }
   }
 
   for (auto &InitCalls : GlobalInitCallMap) {
+    // Don't optimize functions that are marked with the opt.never attribute.
+    bool shouldOptimize = true;
+    for (auto *apply : InitCalls.second) {
+      if (!apply->getFunction()->shouldOptimize()) {
+        shouldOptimize = false;
+        break;
+      }
+    }
+    if (!shouldOptimize)
+      continue;
+
     // Optimize the addressors if possible.
     optimizeInitializer(InitCalls.first, InitCalls.second);
     placeInitializers(InitCalls.first, InitCalls.second);
   }
 
   for (auto &Init : GlobalVarStore) {
+    // Don't optimize functions that are marked with the opt.never attribute.
+    if (!Init.second->getFunction()->shouldOptimize())
+      continue;
+
     // Optimize the access to globals if possible.
     optimizeGlobalAccess(Init.first, Init.second);
   }
 
+  for (auto &addrPair : GlobalAddrMap) {
+    // Don't optimize functions that are marked with the opt.never attribute.
+    bool shouldOptimize = true;
+    for (auto *addr : addrPair.second) {
+      if (!addr->getFunction()->shouldOptimize()) {
+        shouldOptimize = false;
+        break;
+      }
+    }
+    if (!shouldOptimize)
+      continue;
+
+    HasChanged |= tryRemoveGlobalAddr(addrPair.first);
+  }
+
+  for (auto &alloc : AllocGlobalStore) {
+    if (!alloc.second->getFunction()->shouldOptimize())
+      continue;
+    HasChanged |= tryRemoveGlobalAlloc(alloc.first, alloc.second);
+  }
+
+  for (auto &global : Module->getSILGlobals()) {
+    // TODO: For some reaons, if we keep iterating through `getSILGlobals`
```

Comment:
@slavapestov any ideas about the best way to resolve this?

Reply:
Sounds like an iterator invalidation issue, because you're modifying Module->getSILGlobals() while iterating over it. The idiomatic solution in cases like this is to put the globals that will be removed into a temporary vector, and then iterate over that vector and remove each one.

Reply:
Yep, that fixed it, thanks!